### PR TITLE
Ensured that MockitoJUnitRunner is thread safe wrt unused stubs detection

### LINKED ISSUE #332: Finalize Mockito Runners
Currently, we have a couple of Mockito Runners that are a part of public API but they are largely unknown:
 - ConsoleSpammingMockitoJUnitRunner
 - VerboseMockitoJUnitRunner

They attempt to improve the testing / TDD experience by adding warnings for following scenarios:
     - a) unused stubbing
     - b) stubbed method called with different args (and a special variant of a)

Example of a warning:

----

 *** Stubbing warnings from Mockito: ***

stubbed with those args here   -> at org.mockitousage.junitrunner.ModellingVerb
BUT called with different args -> at org.mockitousage.junitrunner.ModellingVerboseMockitoTest.shouldLogUnusedS
This stubbing was never used   -> at org.mockitousage.junitrunner.ModellingVerboseMockitoTest.shouldLogUnusedStubbingWarning
This stubbing was never used   -> at org.mockitousage.junitrunner.ModellingVerboseMockitoTest.shouldLogUnusedStubbingWarningWh

----

The rationale is explained in the javadoc of the above runners. In 2.0 let’s include stubbing warnings in the default behavior of JUnit rule and the default runner.

Challenges that need thinking through:
 - feature parity with JUnit rule
 - clean API so that the warning system can be integrated with other runners
 - some API to get hold of those warnings programmatically
 - feature toggle
     - disable completely
     - disable by test / test class / stubbing?
 - how can this be tested?

### LINKED ISSUE #401: JUnit runner detects unused stubs
It is useful to know that stubbing in the test was not used. Perhaps the stubbing is not needed? Removal of unnecessary code from the test is important for code clarity.

More details:
- we detect unused stubs only for passing tests (adding different failure to an existing failure could be confusing)
- need to have something similar for JUnit rules (will create separate ticket)
- stubbing in @Before or class constructor should be used in at least one test
- stubbing in test needs to be used in that test

### PR DESCRIPTION
**Pull request by szczepiq**
This PR concludes #401 Yay!!!!

Highlights:
- fixed the thread safety problem by leveraging the same API that JUnit rules used for hinting at unused stubs
- removed StubbingListener API (it could not work to the thread safety levels we support in Mockito, see also #545)

Closes #332
Fixes #401

### CODE REVIEW

**Comment by bric3** [comment]
LGTM

**Comment by szczepiq** [comment]
Thanks for review!!! How come you didn't merge yourself?

**Comment by TimvdLippe** [comment]
I still want to take a look at this PR. Can do this evening

**Comment by szczepiq** [comment]
@TimvdLippe, I almost merged it when I saw your reply. Literally I was crafting the squashed commit message :) Anyhow, if you're squashing you can use this summary:

- Fixes #401 - Unused stubbing detection logic was not thread safe in Mockito JUnit runner
- Killed the StubbingListener API - it couldn't support thread safe scenarios Mockito requires
- small javadoc tweaks, removed my login id from javadocs
- deprecated few more APIs connected to the runners
- ensured we verify correctness of MockitoJUnitRunner.Strict (it was not used directly by any of the tests but it was well tested out / great coverage)

File: src/main/java/org/mockito/internal/progress/MockingProgressImpl.java [hunk 1]
```diff
@@ -168,14 +153,14 @@ public VerificationMode maybeVerifyLazily(VerificationMode mode) {
...
 
-    public StubbingListener getStubbingListener() {
-        if (this.stubbingListener == null) {
-            return NoOpStubbingListener.INSTANCE;
-        }
-        return this.stubbingListener;
-    }
+     //TODO 545 thread safety of all mockito
+
+     use cases:
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/progress/MockingProgressImpl.java]: Why does this TODO still exist?

Comment (szczepiq) [inline src/main/java/org/mockito/internal/progress/MockingProgressImpl.java]: yes, see issue #545

**Comment by szczepiq** [comment]
Waiting until EOD today and merging if no other feedback. If you're happy with change merge yourself :)

**Pull request merged**